Stop using c10::scalar_to_tensor in float_power.

gchanan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50115 Stop moving scalars to GPU for one computation in leaky_rrelu_backward.
* #50114 Stop using an unnecessary scalar_to_tensor(..., device) call.
* #50111 Move scalar_to_tensor_default_dtype out of ScalarOps.h because it's only useful for torch.where.
* **#50105 Stop using c10::scalar_to_tensor in float_power.**

There should be no functional change here.

A couple of reasons here:
1) This function is generally an anti-pattern (https://github.com/pytorch/pytorch/issues/49758) and it is good to minimize its usage in the code base.
2) pow itself has a fair amount of smarts like not broadcasting scalar/tensor combinations and we should defer to it.

Differential Revision: [D25786172](https://our.internmc.facebook.com/intern/diff/D25786172)

gchanan committed:
Stop using c10::scalar_to_tensor in float_power.

There should be no functional change here.

A couple of reasons here:
1) This function is generally an anti-pattern (https://github.com/pytorch/pytorch/issues/49758) and it is good to minimize its usage in the code base.
2) pow itself has a fair amount of smarts like not broadcasting scalar/tensor combinations and we should defer to it.

gchanan committed:
Update on "Stop using c10::scalar_to_tensor in float_power."


There should be no functional change here.

A couple of reasons here:
1) This function is generally an anti-pattern (https://github.com/pytorch/pytorch/issues/49758) and it is good to minimize its usage in the code base.
2) pow itself has a fair amount of smarts like not broadcasting scalar/tensor combinations and we should defer to it.

Differential Revision: [D25786172](https://our.internmc.facebook.com/intern/diff/D25786172)

gchanan committed:
Update on "Stop using c10::scalar_to_tensor in float_power."


There should be no functional change here.

A couple of reasons here:
1) This function is generally an anti-pattern (https://github.com/pytorch/pytorch/issues/49758) and it is good to minimize its usage in the code base.
2) pow itself has a fair amount of smarts like not broadcasting scalar/tensor combinations and we should defer to it.

Differential Revision: [D25786172](https://our.internmc.facebook.com/intern/diff/D25786172)

mruberry commented on aten/src/ATen/native/Pow.cpp:
```diff
@@ -89,24 +89,32 @@ Tensor& float_power_out(Tensor& result, const Tensor& base, const Tensor& exp) {
 Tensor& float_power_out(Tensor& result, const Tensor& base, Scalar exp) {
-  return at::float_power_out(result, base, c10::scalar_to_tensor(exp, base.device()));
+  auto dtype = (at::isComplexType(base.scalar_type()) || exp.isComplex()) ? at::kComplexDouble : at::kDouble;
+  exp = dtype == at::kComplexDouble ? exp.toComplexDouble() : exp.toDouble();
```
Parens after the "=" might clarify this expression

mruberry commented on aten/src/ATen/native/Pow.cpp:
```diff
@@ -118,7 +126,11 @@ Tensor& float_power_(Tensor& base, const Tensor& exp) {
-  return base.float_power_(c10::scalar_to_tensor(exp, base.device()));
+  auto dtype = (at::isComplexType(base.scalar_type()) || exp.isComplex()) ? at::kComplexDouble : at::kDouble;
+  TORCH_CHECK(base.scalar_type() == dtype,
+              "self tensor type ", base.scalar_type(), "is not the desired type ", dtype);
```
We're inconsistent in our checks (note: we should pick a style an be consistent) but typically want to mention the operation and use the name in the docs, not "self" which is typically an internal name. Also, "tensor type" is ambiguous and this should refer to dtype.

Maybe an error message like:

"the base given to float_power_ has dtype DTYPE, but the operation's result requires dtype DTYPE"

?

mruberry replied:
Note also there's not a space between the first `base.scalar_type()` and `"is` in the string. A similar printing error can be seen in the error messages below.

gchanan replied:
is this even correct behavior?  Don't we generally check that the computation type is castable to the result type, not that they are the same?

mruberry replied:
Good point. I extended the Developer FAQ entry to address this explicitly (see [here](https://github.com/pytorch/pytorch/wiki/Developer-FAQ#how-does-out-work-in-pytorch)). 

So yes, a corrected version of the check would validate that the operation's common dtype is "safe castable" to base's dtype.

mruberry replied:
In this case, delegating to the out= variant of pow here would probably save adding special logic for this case. That logic needs to be added to the out= variant (per other comments), anyway.

gchanan replied:
Ok, since this is a different issue, I filed https://github.com/pytorch/pytorch/issues/50213.

mruberry commented:
While many of the failures here are in upstream, some are real:

```
test_float_power_constant_cpu - TestAutogradDeviceTypeCPU
test_autograd.py

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 284, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 279, in instantiated_test
    result = test_fn(self, *args)
  File "test_autograd.py", line 5206, in do_test
    check(name)
  File "test_autograd.py", line 5177, in check
    **kwargs_variable))
RuntimeError: result type ComplexDoublecan't be cast to the desired output type Double

test_float_power_constant_cuda - TestAutogradDeviceTypeCUDA
test_autograd.py

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 889, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 889, in wrapper
    method(*args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 284, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 279, in instantiated_test
    result = test_fn(self, *args)
  File "test_autograd.py", line 5206, in do_test
    check(name)
  File "test_autograd.py", line 5177, in check
    **kwargs_variable))
RuntimeError: result type ComplexDoublecan't be cast to the desired output type Double

test_float_power_constant_cpu - test_jit.TestJitGeneratedAutogradCPU
test_jit.py

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 284, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 279, in instantiated_test
    result = test_fn(self, *args)
  File "/var/lib/jenkins/workspace/test/test_jit.py", line 15849, in do_test
    check(inplace_name)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/jit_utils.py", line 646, in wrapper
    fn(*args, **kwargs)
  File "/var/lib/jenkins/workspace/test/test_jit.py", line 15842, in check
    check_alias_annotation(name, (self_variable,) + args_variable, kwargs_variable, aten_name=name)
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/jit_metaprogramming_utils.py", line 477, in check_alias_annotation
    torch._C._jit_check_alias_annotation(CU.the_method.graph, tuple(tensors), aten_name)
RuntimeError: result type ComplexDoublecan't be cast to the desired output type Double
```

mruberry commented:
The error message, with its incorrect spacing, comes from:

https://github.com/pytorch/pytorch/blob/97c17b47720dd1afce88180b843c9182f1a665a9/aten/src/ATen/native/Pow.cpp#L30

mruberry commented:
It's not obvious to me where this failure is coming from.

We could change float_power to use an OpInfo if there's a test issue.

mruberry commented on aten/src/ATen/native/Pow.cpp:
```diff
@@ -89,24 +89,32 @@ Tensor& float_power_out(Tensor& result, const Tensor& base, const Tensor& exp) {
-  auto dtype = (at::isComplexType(base.scalar_type()) || at::isComplexType(exp.scalar_type())) ? at::kComplexDouble : at::kDouble;
-  return at::pow(base.to(dtype), exp.to(dtype));
+  auto dtype = (at::isComplexType(exp.scalar_type()) || base.isComplex()) ? at::kComplexDouble : at::kDouble;
+  base = dtype == at::kComplexDouble ? base.toComplexDouble() : base.toDouble();
```
Ideally we'd add an output checking helper (or maybe one exists that we can reuse) here. Otherwise if result has the wrong dtype or the wrong device type the error will be throw in pow, and will not correctly identify float_power as the source of the issue. This could be confusing to users.

gchanan committed:
Update on "Stop using c10::scalar_to_tensor in float_power."


There should be no functional change here.

A couple of reasons here:
1) This function is generally an anti-pattern (https://github.com/pytorch/pytorch/issues/49758) and it is good to minimize its usage in the code base.
2) pow itself has a fair amount of smarts like not broadcasting scalar/tensor combinations and we should defer to it.

Differential Revision: [D25786172](https://our.internmc.facebook.com/intern/diff/D25786172)

gchanan committed:
Update on "Stop using c10::scalar_to_tensor in float_power."


There should be no functional change here.

A couple of reasons here:
1) This function is generally an anti-pattern (https://github.com/pytorch/pytorch/issues/49758) and it is good to minimize its usage in the code base.
2) pow itself has a fair amount of smarts like not broadcasting scalar/tensor combinations and we should defer to it.

Differential Revision: [D25786172](https://our.internmc.facebook.com/intern/diff/D25786172)

gchanan committed:
Update on "Stop using c10::scalar_to_tensor in float_power."


There should be no functional change here.

A couple of reasons here:
1) This function is generally an anti-pattern (https://github.com/pytorch/pytorch/issues/49758) and it is good to minimize its usage in the code base.
2) pow itself has a fair amount of smarts like not broadcasting scalar/tensor combinations and we should defer to it.

Differential Revision: [D25786172](https://our.internmc.facebook.com/intern/diff/D25786172)

mruberry reviewed (approved):
LGTM!
